src/storage/mod.rs

```diff
@@ -812,15 +868,42 @@ impl Storage {
         &self,
         ctx: Context,
         key: Vec<u8>,
-        callback: Callback<Option<Vec<u8>>>,
-    ) -> Result<()> {
-        let cmd = Command::RawGet {
-            ctx: ctx,
-            key: Key::from_encoded(key),
-        };
-        self.schedule(cmd, StorageCb::SingleValue(callback))?;
-        RAWKV_COMMAND_COUNTER_VEC.with_label_values(&["get"]).inc();
-        Ok(())
+    ) -> Box<Future<Item = Option<Vec<u8>>, Error = Error> + Send> {
+        static CMD: &'static str = "raw_get";
+        let engine = self.get_engine();
+        let priority = readpool::Priority::from(ctx.get_priority());
+
+        let res = self.read_pool.future_execute(priority, move |ctxd| {
+            let mut _timer = {
+                let ctxd = ctxd.clone();
+                let mut thread_ctx = ctxd.current_thread_context_mut();
+                thread_ctx.collect_command_count_and_duration(CMD, priority, true)
```

Comment:
Could we define L877-880 into an independent function since it's repeated in each read operation?

Reply:
That will not be an elegant solution, due to:

- `collect_command_count_and_duration` should be kept standalone because it is not related to "get a current context".
- Under the background above, we will only extract Line 878 and Line 879. However `current_thread_context_mut` returns a `RefMut`, which is some kind of guard (needs a longer lifetime) and must be kept in a standalone line if it is extracted into a function.
- We will only save 1 line if it is extracted. It is still a bit verbose.

I think the best solution is to implement some kind of smart-pointer like behaviour (i.e. DeRef) for ctxd in futurepool. However we met some issues when attempting to implement it previously due to a complex life time bound in RefMut. So that this leaves open for now. We can refine in future if someone figures out how to do that. Note, it is a work to be done in futurepool and is not related to this PR as well.

Hope you understand :)

Reply:
Here is a risky but elegant implementation for the proposal above:

```rust

/// A delegator to wrap `Context` to provide `on_tick` feature.
#[derive(Debug)]
struct ContextDelegator<T: Context> {
    inner: UnsafeCell<T>,
    ...
}

impl<T: Context> ContextDelegator<T> {
    /// Ensure that access to this function is unique
    fn context_mut(&self) -> &mut T {
        unsafe { &mut *self.inner.get() }
    }

    ...
}

impl<T: Context> ops::Deref for ContextDelegators<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        self.current_thread_context_mut()
    }
}

impl<T: Context> ops::DerefMut for ContextDelegators<T> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        self.current_thread_context_mut()
    }
}
```

and the code you quoted can be simplified to only 1 line:

```rust
let mut _timer = ctxd.clone().collect_command_count_and_duration(CMD, priority, true);
```

however this solution requires much testing. We can consider it later.